src/chain/tracker.rs

```diff
@@ -116,9 +123,117 @@ pub fn start_chain_watch(
                                     break;
                                 }
 
-                                let mut id_remove_list = Vec::new();
+                                tracing::debug!("Watched accounts: {watched_accounts:?}");
+
                                 let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_millis() as u64;
 
+                                let mut id_remove_list = Vec::new();
+
+                                match transfer_events(
+                                    &client,
+                                    &block,
+                                    &watcher.metadata,
+                                )
+                                    .await {
+                                        Ok((timestamp, events)) => {
+                                        tracing::debug!("Got a block with timestamp {timestamp:?} & events: {events:?}");
+
+                                        for (id, invoice) in &watched_accounts {
+                                            for (extrinsic_option, event) in &events {
+                                                if let Some((tx_kind, another_account, transfer_amount)) = parse_transfer_event(&invoice.address, &event.0.fields) {
+                                                    tracing::debug!("Found {tx_kind:?} from/to {another_account:?} with {transfer_amount:?} token(s).");
+
+                                                    let Some((position_in_block, extrinsic)) = extrinsic_option else {
+                                                        return Err(Error::from(ChainError::TransferEventNoExtrinsic));
+                                                    };
+
+                                                    let finalized_tx_timestamp = (OffsetDateTime::UNIX_EPOCH + Duration::from_millis(timestamp.0))
+                                                        .format(&Rfc3339).unwrap().into();
+                                                    let finalized_tx = FinalizedTxDb {
+                                                            block_number,
+                                                            position_in_block: *position_in_block
+                                                        }.into();
+                                                    let amount = Amount::Exact(transfer_amount.format(invoice.currency.decimals));
+                                                    let status = TxStatus::Finalized;
+                                                    let currency = invoice.currency.clone();
+                                                    let transaction_bytes = const_hex::encode_prefixed(extrinsic);
+
+                                                    match tx_kind {
+                                                        kind @ TxKind::Payment => {
```

Comment:
Was just about to merge this, but realized - this might misfire. If there is an eventful transfer and eventless NEW AWESOME FEATURE that will reap some cash away somehow, we'll end up with undefined state where payment is recognized, but account balance is below expected. Let's remove this variant's marking of account as paid since we do it below with more robust check?

Reply:
https://github.com/Alzymologist/Kalatori-backend/pull/107/commits/558eb7a2068dd02eb571dd73b83ec2d5c290bbcf